### PR TITLE
Fix S4433 FP

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes: #2537

### CODE REVIEW

**Comment by andrei-epure-sonarsource** [comment]
related to #2537

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs [hunk 1]
```diff
@@ -216,7 +216,7 @@ private static bool IsAllowedValueForXmlTextReader(object constantValue)
                     symbol.Kind == SymbolKind.Method &&
                     symbol.ContainingType.GetSymbolType().IsAny(UnsafeXmlResolvers);
 
-            private static bool IsAllowedObject(ISymbol symbol) =>
+            private static bool IsAllowedObject(ISymbol symbol, SyntaxNode node, SemanticModel model) =>
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs]: I'd comment that we don't use `node` and `model` on purpose (to avoid confusion in the future)

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LdapConnectionShouldBeSecure.cs [hunk 2]
```diff
@@ -129,4 +129,64 @@ void Cases()
             Property6 = new DirectoryEntry(); // Compliant
         }
     }
+
+    public class AllTypes
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LdapConnectionShouldBeSecure.cs]: can you also add test cases (unless I miss them) for initializations inside
- constructor
- inside getter / setter (for properties we only have `private DirectoryEntry Property3 { get; set; } = new DirectoryEntry()...`)
- struct
- interface default method

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs [hunk 3]
```diff
@@ -43,14 +47,64 @@ public sealed class LdapConnectionShouldBeSecure : ObjectShouldBeInitializedCorr
...
 
         private static bool IsAllowedValue(object constantValue) =>
             constantValue is int integerValue &&
-            (integerValue & AuthenticationTypes_Secure) > 0; // The expected value is a bit from a Flags enum
-
+            IsValidAuthenticationType(integerValue);
 
         private static readonly ImmutableArray<KnownType> TrackedTypes = ImmutableArray.Create(KnownType.System_DirectoryServices_DirectoryEntry);
+
+        private static bool IsAllowedObject(ISymbol symbol, SyntaxNode node, SemanticModel semanticModel)
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs]: could you add a comment for this rule what is the symbol, what is the node ?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs [hunk 4]
```diff
@@ -43,14 +47,64 @@ public sealed class LdapConnectionShouldBeSecure : ObjectShouldBeInitializedCorr
...
+            root.DescendantNodes()
+                .OfType<VariableDeclaratorSyntax>()
+                .Where(variableDeclarator => VariableDeclaratorHasSymbol(variableDeclarator, symbol, semanticModel))
+                .Any(variableDeclarator => HasUnsafeConstantValue(variableDeclarator.Initializer.Value, semanticModel));
+
+        private static bool VariableDeclaratorHasSymbol(VariableDeclaratorSyntax variableDeclarator, ISymbol symbol, SemanticModel semanticModel) =>
+            variableDeclarator.Identifier.ValueText == symbol.Name &&
+            symbol.Equals(semanticModel.GetDeclaredSymbol(variableDeclarator));
+
+        private static bool HasUnsafeConstantValue(SyntaxNode node, SemanticModel semanticModel) =>
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs]: the logic is a bit hard to follow
- HasUnsafeConstantValue - negative in name
- `!IsValidAuthenticationType(authType)` - negation with positive name
- `IsValidAuthenticationType` contains negations inside `authType != AuthenticationTypesNone && authType != AuthenticationTypesAnonymous`

negative logic is harder to follow

if would be easier to read 
```
private static bool HasUnsafeConstantValue(SyntaxNode node, SemanticModel semanticModel) =>
 // ...
 // IsUnsafeAuthenticationType(authType);	 
```

and then the only negation would be in the names , and would be consistent

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs [hunk 5]
```diff
@@ -43,14 +47,64 @@ public sealed class LdapConnectionShouldBeSecure : ObjectShouldBeInitializedCorr
...
+            // In order to correctly detect all the cases we will need to refactor the rule and use symbolic execution.
+            // Until this is done, we can reduce the number of false positives by checking if the variable has one of the following
+            // values assigned in the parent scope: AuthenticationTypes.None, AuthenticationTypes.Anonymous, default or default(AuthenticationTypes).
+            var root = node.FirstAncestorOrSelf<MethodDeclarationSyntax>() ??
+                       node.FirstAncestorOrSelf<ClassDeclarationSyntax>() ??
+                       node.FirstAncestorOrSelf<StructDeclarationSyntax>() ??
+                       node.FirstAncestorOrSelf<InterfaceDeclarationSyntax>() ??
+                       node.SyntaxTree.GetRoot();
+
+            return !HasUnsafeDeclaration(root, symbol, semanticModel) &&
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs]: it would be nice to add a comment specifying what these two checks do, the scenario(s), because it's not obvious and the methods themselves don't help either

and basically explain how it's complementary to what the ObjectInitializationTracker verifies

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LdapConnectionShouldBeSecure.cs [hunk 6]
```diff
@@ -129,4 +129,64 @@ void Cases()
...
+        {
+            var authType = condition ? AuthenticationTypes.SecureSocketsLayer : AuthenticationTypes.Secure;
+            var entry = new DirectoryEntry("", null, null, authType);
+        }
+
+        public void WithTernaryOperatorUnsecure(bool condition)
+        {
+            var authType = condition ? AuthenticationTypes.None : AuthenticationTypes.Anonymous;
+            // Symbolic execution would a better fit to increase precision
+            var entry = new DirectoryEntry("", null, null, authType); // Compliant - FN
```
[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LdapConnectionShouldBeSecure.cs]: as discussed, I will not open a new issue since this is by design (the rule does not use symbolic execution) and this kind of false negatives is expected.
Indeed we need to make this more visible to the AppSec team but this will be handled separately.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs [hunk 7]
```diff
@@ -53,28 +53,37 @@ public sealed class LdapConnectionShouldBeSecure : ObjectShouldBeInitializedCorr
...
-
-            return !HasUnsafeDeclaration(root, symbol, semanticModel) &&
-                   !HasUnsafeAssignment(root, symbol, semanticModel);
+            var root = authTypeExpression.FirstAncestorOrSelf<MethodDeclarationSyntax>() ??
+                       authTypeExpression.FirstAncestorOrSelf<ClassDeclarationSyntax>() ??
+                       authTypeExpression.FirstAncestorOrSelf<StructDeclarationSyntax>() ??
+                       authTypeExpression.FirstAncestorOrSelf<InterfaceDeclarationSyntax>() ??
+                       authTypeExpression.SyntaxTree.GetRoot();
+
+            // The CSharpObjectInitializationTracker is verifying only if AuthenticationType property or DirectoryEntry ctor
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs]:
```suggestion
            // The CSharpObjectInitializationTracker is verifying only if the AuthenticationType property or DirectoryEntry ctor
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs [hunk 8]
```diff
@@ -53,28 +53,37 @@ public sealed class LdapConnectionShouldBeSecure : ObjectShouldBeInitializedCorr
...
+            // was invoked with a valid value, which can be evaluated as a constant.
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs]:
```suggestion
            // was invoked with a valid constant value.
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs [hunk 9]
```diff
@@ -53,28 +53,37 @@ public sealed class LdapConnectionShouldBeSecure : ObjectShouldBeInitializedCorr
...
+            // If the parameter is a variable, in order to reduce FP, we have to check if an unsafe value was set during
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LdapConnectionShouldBeSecure.cs]:
```suggestion
            // When the initialization is done with a variable, we check if an unsafe value was set during
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs [hunk 10]
```diff
@@ -173,15 +174,15 @@ public static (CSharpObjectInitializationTracker xmlDocumentTracker, CSharpObjec
                     isAllowedConstantValue: constantValue => false,
                     trackedTypes: XmlDocumentTrackedTypes,
                     isTrackedPropertyName: propertyName => "XmlResolver" == propertyName,
-                    isAllowedObject: IsAllowedObject,
+                    isAllowedObject: (symbol, _, __) => IsAllowedObject(symbol),
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs]: is it me or are there two underscores `__` ?
is it intentional?
(also see line 185)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SyntaxTrackers/CSharpObjectInitializationTracker.cs [hunk 11]
```diff
@@ -56,9 +56,9 @@ public class CSharpObjectInitializationTracker
         private readonly Predicate<object> isAllowedConstantValue;
 
         /// <summary>
-        /// Given the symbol of an object, analyzed syntaxNode and the semantic model, returns true if it is allowed.
+        /// Given the symbol of an object, expresion used to populate the value and the semantic model, returns true if it is allowed.
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SyntaxTrackers/CSharpObjectInitializationTracker.cs]:
```suggestion
        /// Given the symbol of an object, the expression used to populate the value and the semantic model, returns true if it is allowed.
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs (hunk 10 above)
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs]: Indeed there are two underscores and it was intentional. AFAICT it's not possible to use _ for both parameters. The compiler will complain that the parameter name is a duplicate.
